File: chess/move.cpp
/* Jerry - A Chess Graphical User Interface
 * Copyright (C) 2014-2016 Dominik Klein
 * Copyright (C) 2015-2016 Karl Josef Klein
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */


#include "move.h"
#include <assert.h>
#include <iostream>
#include <QDebug>

namespace chess {

Move::Move(uint8_t from, uint8_t to) {

    // create ascii (latin1) code numbers from
    // uint8_t board pos numbers
    QChar col_from = QChar((from % 10) + 96);
    QChar row_from = QChar((from / 10) + 47);

    QChar col_to = QChar((to % 10) + 96);
    QChar row_to = QChar((to / 10) + 47);

    this->from = from;
    this->to = to;
    this->promotion_piece = 0;
    this->uci_string = QString(col_from) + row_from + col_to + row_to;
    this->is_null = false;
}

/*
 * creates a null move
 */
Move::Move() {
    this->from = 0x00;
    this->to = 0x00;
    this->promotion_piece = 0;
    this->uci_string = "0000";
    this->is_null = true;
}

Move::Move(uint8_t from, uint8_t to, uint8_t promotion_piece) {
    QChar col_from = QChar((from % 10) + 96);
    QChar row_from = QChar((from / 10) + 47);

    QChar col_to = QChar((to % 10) + 96);
    QChar row_to = QChar((to / 10) + 47);

    QChar prom_piece;
    if(promotion_piece == KNIGHT) {
        prom_piece = QChar('N');
    } else if (promotion_piece == BISHOP) {
        prom_piece = QChar('B');
    } else if (promotion_piece == ROOK) {
        prom_piece = QChar('R');
    } else if (promotion_piece == QUEEN) {
        prom_piece = QChar('Q');
    } else {
        prom_piece = QChar(' ');
    }
    this->from = from;
    this->to = to;
    this->promotion_piece = promotion_piece;
    if(prom_piece == QChar(' ')) {
        this->uci_string = QString(col_from) + row_from + col_to + row_to;
    } else {
        this->uci_string = QString(col_from) + row_from + col_to + row_to + prom_piece;
    }
    this->is_null = false;
}

Move::Move(QString uci) {
    assert((uci.size()==4) || (uci.size()==5));
    this->uci_string = uci;
    QString up = uci.toUpper();
    uint8_t from_col = this->alpha_to_pos(up.at(0));
    // -49 for ascii(1) -> int 0, *10 + 20 is to get board coord
    uint8_t from_row = (uint8_t) ((up.at(1).toLatin1()-49) * 10)+20;
    this->from = from_row + from_col;
    uint8_t to_col = this->alpha_to_pos(up.at(2));
    uint8_t to_row = (uint8_t) ((up.at(3).toLatin1() -49) * 10) + 20;
    this->to = to_row + to_col;
    if(uci.size() == 5) {
        QChar piece = up.at(4);
        assert(piece == QChar('N') || piece == QChar('B') ||
               piece == QChar('R') || piece == QChar('Q'));
        if(piece == QChar('N')) {
            this->promotion_piece = KNIGHT;
        }
        if(piece == QChar('B')) {
            this->promotion_piece = BISHOP;
        }
        if(piece == QChar('R')) {
            this->promotion_piece = ROOK;
        }
        if(piece == QChar('Q')) {
            this->promotion_piece = QUEEN;
        }
    } else {
        this->promotion_piece = 0;
    }
    this->is_null = false;
}

QString Move::uci() {
    if(this->is_null) {
        return "0000";
    } else {
        //return this->uci_string;
        QChar col_from = QChar((this->from % 10) + 96);
        QChar row_from = QChar((this->from / 10) + 47);

        QChar col_to = QChar((this->to % 10) + 96);
        QChar row_to = QChar((this->to / 10) + 47);

        QString uci = QString(col_from) + row_from + col_to + row_to;
        if(this->promotion_piece==BISHOP) {
            uci.append("=B");
        } else if(this->promotion_piece==KNIGHT) {
            uci.append("=K");
        } else if(this->promotion_piece==ROOK) {
            uci.append("=R");
        } else if(this->promotion_piece==QUEEN) {
            uci.append("=Q");
        } else {
            return uci;
        }
        return uci;
    }
}

QPoint Move::fromAsXY() {
    int col_from = (this->from % 10) - 1;
    int row_from = (this->from / 10) - 2;
    return QPoint(col_from, row_from);
}

QPoint Move::toAsXY() {
    int col_to = (this->to % 10) - 1;
    int row_to = (this->to / 10) - 2;
    return QPoint(col_to, row_to);
}

// create deep copy
Move::Move(const Move& m) {
    this->from = m.from;
    this->to = m.to;
    this->promotion_piece = m.promotion_piece;
    this->is_null = m.is_null;
    this->uci_string = this->uci(); // QString::fromUtf16(m.uci_string.utf16());
}


uint8_t Move::alpha_to_pos(QChar alpha) {
    if(alpha == QChar('A')) {
        return 1;
    } else if(alpha == QChar('B')) {
        return 2;
    } else if(alpha == QChar('C')) {
        return 3;
    } else if(alpha == QChar('D')) {
        return 4;
    } else if(alpha == QChar('E')) {
        return 5;
    } else if(alpha == QChar('F')) {
        return 6;
    } else if(alpha == QChar('G')) {
        return 7;
    } else if(alpha == QChar('H')) {
        return 8;
    }

    return 0;
}

bool Move::operator==(const Move &other) const {
    return (this->from == other.from
            && this->to == other.to
            && this->promotion_piece == other.promotion_piece
            && this->is_null == other.is_null);
}

bool Move::operator!=(const Move &other) const {
  return !(*this == other);
}

/**
 * @brief operator <<
 * @param strm
 * @param m
 * @return
 *
 * prints move as uci string
 *
 * Example:
 * Move m = Move("g1f3");
 * std::cout << m << std::endl;
 */
std::ostream& operator<<(std::ostream &strm, const Move &m) {

    return strm << m.uci_string.toStdString();

}

}

File: src/main/java/org/asdfjkl/jerryfx/lib/Move.java
/* JerryFX - A Chess Graphical User Interface
 * Copyright (C) 2020 Dominik Klein
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.asdfjkl.jerryfx.lib;

public class Move {

    int from;
    int to;
    int promotionPiece;
    boolean isNullMove;

    public Move(int from, int to) {

        this.from = from;
        this.to = to;
        this.promotionPiece = 0;
        this.isNullMove = false;

    }

    public Move(int fromColumn, int fromRow, int toColumn, int toRow) {

        this.from = ((fromRow + 2) * 10) + (fromColumn + 1);
        this.to = ((toRow + 2) * 10) + (toColumn + 1);
        this.promotionPiece = 0;
        this.isNullMove = false;

    }

    public Move() {

        this.from = 0;
        this.to = 0;
        this.promotionPiece = 0;
        this.isNullMove = true;

    }

    public Move(int from, int to, int promotionPiece) {

        this.from = from;
        this.to = to;
        this.promotionPiece = promotionPiece;

    }

    public Move(int fromColumn, int fromRow, int toColumn, int toRow, char promotionPiece) {

        //System.out.println("from col: "+fromColumn + " tocol: "+toColumn);
        //System.out.println("from row: "+fromRow + " torow: "+toRow);
        this.from = ((fromRow + 2) * 10) + (fromColumn + 1);
        this.to = ((toRow + 2) * 10) + (toColumn + 1);

        //System.out.println("intfrom: "+this.from);
        //System.out.println("intto: "+this.to);


        this.promotionPiece = -1;
        if (promotionPiece == 'N') {
            this.promotionPiece = CONSTANTS.KNIGHT;
        }
        if (promotionPiece == 'B') {
            this.promotionPiece = CONSTANTS.BISHOP;
        }
        if (promotionPiece == 'R') {
            this.promotionPiece = CONSTANTS.ROOK;
        }
        if (promotionPiece == 'Q') {
            this.promotionPiece = CONSTANTS.QUEEN;
        }
        if (this.promotionPiece < 0) {
            throw new IllegalArgumentException("Illegal Promotion Piece: " + promotionPiece);
        }
        this.isNullMove = false;

    }

    public void setPromotionPiece(int promotionPiece) {
        this.promotionPiece = promotionPiece;
    }

    private int alphaToPos(char alpha) {
        if (alpha == 'A') {
            return 1;
        } else if (alpha == 'B') {
            return 2;
        } else if (alpha == 'C') {
            return 3;
        } else if (alpha == 'D') {
            return 4;
        } else if (alpha == 'E') {
            return 5;
        } else if (alpha == 'F') {
            return 6;
        } else if (alpha == 'G') {
            return 7;
        } else if (alpha == 'H') {
            return 8;
        }
        throw new IllegalArgumentException("alpha to pos called with: " + alpha);
    }


    public Move(String uci) {

        if (!(uci.length() == 4 || uci.length() == 5)) {
            throw new IllegalArgumentException("Illegal Uci String: " + uci);
        }
        String uciUpper = uci.toUpperCase();

        int fromColumn = this.alphaToPos(uciUpper.charAt(0));
        // -49 is for ascii(1) -> int 0, * 10 + 20 is to get internal board coordinate
        int fromRow = ((((int) uciUpper.charAt(1)) - 49) * 10) + 20;
        this.from = fromRow + fromColumn;

        int toColumn = this.alphaToPos(uciUpper.charAt(2));
        int toRow = ((((int) uciUpper.charAt(3)) - 49) * 10) + 20;
        this.to = toRow + toColumn;

        if (uciUpper.length() == 5) {
            char promPiece = uciUpper.charAt(4);
            this.promotionPiece = -1;
            if (promPiece == 'N') {
                this.promotionPiece = CONSTANTS.KNIGHT;
            }
            if (promPiece == 'B') {
                this.promotionPiece = CONSTANTS.BISHOP;
            }
            if (promPiece == 'R') {
                this.promotionPiece = CONSTANTS.ROOK;
            }
            if (promPiece == 'Q') {
                this.promotionPiece = CONSTANTS.QUEEN;
            }
            if (promotionPiece < 0) {
                throw new IllegalArgumentException("illegal uci string: " + uci);
            }
        }
        this.isNullMove = false;

    }

    public String getUci() {

        if (this.isNullMove) {
            return "0000";
        } else {
            char colFrom = (char) ((this.from % 10) + 96);
            char rowFrom = (char) ((this.from / 10) + 47);

            //System.out.println(("from: "+this.from));

            char colTo = (char) ((this.to % 10) + 96);
            char rowTo = (char) ((this.to / 10) + 47);

            //System.out.println(("to: "+this.to));

            String uci = "";
            uci += colFrom;
            uci += rowFrom;
            uci += colTo;
            uci += rowTo;
            if (this.promotionPiece == CONSTANTS.KNIGHT) {
                uci += "N";
            }
            if (this.promotionPiece == CONSTANTS.ROOK) {
                uci += "R";
            }
            if (this.promotionPiece == CONSTANTS.QUEEN) {
                uci += "Q";
            }
            if (this.promotionPiece == CONSTANTS.BISHOP) {
                uci += "B";
            }
            return uci;
        }
    }

    @Override
    public String toString() {
        return this.getUci();
    }


}